futures-core/src/task_impl/mod.rs

```diff
@@ -217,42 +187,28 @@ impl<T: ?Sized> Spawn<T> {
         where N: Clone + Into<NotifyHandle>,
               T: Stream,
     {
-        self.poll_notify(notify, id, |s| s.poll())
+        self.poll_notify(notify, id, |s, ctx| s.poll(ctx))
     }
 
     /// Invokes the function with the provided `notify` in the task context.
     pub fn poll_notify<N, F, R>(&mut self, notify: &N, id: usize, f: F) -> R
         where N: Clone + Into<NotifyHandle>,
-              F: FnOnce(&mut T) -> R
+              F: FnOnce(&mut T, &mut Context) -> R
     {
         let mk = || notify.clone().into();
         self.enter(BorrowedUnpark::new(&mk, id), f)
     }
 
-    /// TODO: dox
-    pub fn with_task_data<F, R>(&mut self, f: F) -> R
-        where F: FnOnce(&mut T) -> R,
-    {
-        let Spawn { ref data, ref mut obj, .. } = *self;
-        with(|task| {
-            let new_task = BorrowedTask {
-                unpark: task.unpark,
-                map: data,
-            };
-
-            set(&new_task, || f(obj))
-        })
-    }
-
     fn enter<F, R>(&mut self, unpark: BorrowedUnpark, f: F) -> R
```

Comment:
The whole `Spawn` type I think may not make 100% sense in the world where contexts are explicitly passed around. I'd sort of naively expect that we'd have something like:

```rust
struct Spawn<F> {
    inner: F,
    context: Context,
}
```

but that unfortunately doesn't work well with the lifetime argument on `Context` right now. In theory though the task-local data should be stored on the `Context` itself for now rather than in `Spawn` I think?

I wonder if we could do something like:

```rust
struct Context {
    data: TaskLocalDataMap, // I forget the actual name
    waker: NotifyHandle, // how to do a wakeup
}

impl Context {
    fn new<T: Into<NotifyHandle>>(inner: T) -> Context;
   
    // variuos accessors for `data`

    fn set_waker(&mut self, ...); // maybe?
}
```

One problem I think is that right now we make heavy usage of the lifetime parameters and stack closures and whatnot. In that sense we sort of want there to be a `Context` type (with no lifetime parameters maybe?) as well as a `ContextRef<'a>` type which all futures take. Or at least that matches that current scheme... I'm sort of losing myself here anyway

Reply:
I did try an iteration of this code with `context` as a member of `Spawn`. It makes panic safety harder when trying to change the notify handle temporarily for the task because you have to have a way to unset it. But you need mut access to the future and the context at the same time (which rust doesn't like).